### PR TITLE
Run tests on older JVMs to check version-specific logic.

### PR DESCRIPTION
**Pull request by anuraaga**
We have more and more version-specific logic now so we should make sure to run tests on multiple versions.

### CODE REVIEW

File: .travis.install_jdk.sh [hunk 1]
```diff
@@ -7,8 +7,10 @@
 
 wget https://github.com/sormuras/bach/raw/master/install-jdk.sh || exit $?
 
-if [ ! -z "$1" ]; then
-  source ./install-jdk.sh --feature $1
+if [ -n "$1" ]; then
```
Comment (trustin) [inline .travis.install_jdk.sh]: `[[ .. ]]` is better than `[ .. ]` :smile:

File: .travis.install_jdk.sh [hunk 2]
```diff
@@ -7,8 +7,10 @@
...
+  # Can't source twice from install-jdk.sh in same build
+  export JAVA_HOME=$HOME/openjdk13
```
Comment (trustin) [inline .travis.install_jdk.sh]: `"$HOME/openjdk13"` (quotation)

File: .travis.install_jdk.sh [hunk 3]
```diff
@@ -7,8 +7,10 @@
...
+  ./install-jdk.sh --feature $1 --target $JAVA_HOME
```
Comment (trustin) [inline .travis.install_jdk.sh]: `"$1"` and `"$JAVA_HOME"` (quotation)

File: .travis.install_jdk.sh [hunk 4]
```diff
@@ -7,8 +7,10 @@
...
 fi
 
 export JAVA_TEST_HOME=$JAVA_HOME
```
Comment (trustin) [inline .travis.install_jdk.sh]: `"$JAVA_HOME"` (quotation)

**Comment by anuraaga** [comment]
Hmm getting very confusing issues since installing openjdk13 will keep on looking

**Comment by trustin** [comment]
- Class version errors on Java 13 builds
- Issues related with ALPN on Java 8 builds (I guess we need to enable Jetty ALPN agent if we didn't?)

**Comment by anuraaga** [comment]
The intent is to continue building on Java 11 but for some reason my Java home isn't working right, will see what's up.

File: .travis.install_jdk.sh [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+# For some reason, official Travis CI docs for using install-jdk.sh don't work so we use a different one.
+# https://docs.travis-ci.com/user/languages/java/#switching-jdks-to-java-10-and-up-within-one-job
+#
+# https://github.com/sormuras/sormuras.github.io/blob/master/.travis.yml
+
+wget https://github.com/sormuras/bach/raw/master/install-jdk.sh
+
+if [[ -n "$1" ]]; then
+  FEATURE="$1"
+  shift 1
```
Comment (trustin) [inline .travis.install_jdk.sh]: nit: You can omit 1.

File: .travis.yml [hunk 6]
```diff
@@ -19,16 +28,14 @@ cache:
...
-- "./gradlew --version"
 install:
 - true
 script:
+- "./gradlew --version"
 - "./gradlew --no-daemon --stacktrace --warning-mode=all -Pcoverage checkstyle check build"
 before_cache:
 - find $HOME/.gradle/caches -name '*.lock' -delete
-after_success:
-- bash <(curl -s https://codecov.io/bash)
```
Comment (trustin) [inline .travis.yml]: Are you going to re-enable this later, or only for a certain Java version?

Comment (anuraaga) [reply .travis.yml]: I figured we should upload coverage once per build. Do you think it's good to upload for all of them?

Comment (trustin) [reply .travis.yml]: I guess we could just pick and upload only one? (Could disable coverage for others for faster builds)

Comment (anuraaga) [reply .travis.yml]: Yeah - I do that above in the matrix :) But good idea to disable coverage on the other builds

**Comment by anuraaga** [comment]
Finally ready!

File: gradle/scripts/lib/java-alpn.gradle [hunk 7]
```diff
@@ -1,5 +1,5 @@
 // JDK 9 does not require Jetty ALPN agent at all.
-if (JavaVersion.current() >= JavaVersion.VERSION_1_9) {
+if (JavaVersion.current() >= JavaVersion.VERSION_1_9 && System.env.TEST_JAVA_VERSION != '8') {
```
Comment (trustin) [inline gradle/scripts/lib/java-alpn.gradle]: I have updated `gradle-scripts` for you. You'll not need this anymore once you rebase.
https://github.com/line/gradle-scripts/commit/a0a57120fb832fe93d23ce335d7316e99b756077

File: core/src/test/java9/com/linecorp/armeria/internal/RequestContextAwareCompletableFutureTest.java [hunk 8]
```diff
@@ -25,15 +26,22 @@
...
 
 import com.linecorp.armeria.common.HttpMethod;
 import com.linecorp.armeria.common.HttpRequest;
 import com.linecorp.armeria.common.RequestContext;
+import com.linecorp.armeria.common.util.SystemInfo;
 import com.linecorp.armeria.server.ServiceRequestContextBuilder;
 
 class RequestContextAwareCompletableFutureTest {
 
+    @BeforeEach
```
Comment (trustin) [inline core/src/test/java9/com/linecorp/armeria/internal/RequestContextAwareCompletableFutureTest.java]: Doesn't `@BeforeAll` work?

Comment (anuraaga) [reply core/src/test/java9/com/linecorp/armeria/internal/RequestContextAwareCompletableFutureTest.java]: Oops yeah

**Comment by trustin** [comment]
Oops. Checkstyle violation:

```
RequestContextAwareCompletableFutureTest.java:30:8: Unused import - org.junit.jupiter.api.BeforeEach. [UnusedImports]
```

File: core/src/test/java9/com/linecorp/armeria/internal/RequestContextAwareCompletableFutureTest.java [hunk 9]
```diff
@@ -25,15 +26,22 @@
...
 import com.linecorp.armeria.common.HttpRequest;
 import com.linecorp.armeria.common.RequestContext;
+import com.linecorp.armeria.common.util.SystemInfo;
 import com.linecorp.armeria.server.ServiceRequestContextBuilder;
 
 class RequestContextAwareCompletableFutureTest {
 
+    @BeforeAll
+    static void checkEnv() {
+        assumeThat(SystemInfo.javaVersion()).isGreaterThanOrEqualTo(9);
```
Comment (ikhoon) [inline core/src/test/java9/com/linecorp/armeria/internal/RequestContextAwareCompletableFutureTest.java]: 👍

**Pull request merged**